### PR TITLE
feat: support addon index

### PR DESCRIPTION
**Pull request by 1aal**
- fix: #34 

kbcli addon index support:

1. add: add a new addon index eg: `kbcli addon index add kubeblocks https://github.com/apecloud/block-index.git`
2. delete: delete an addon index eg: `kbcli addon index delete kubeblocks`
3. update: update the specified index(es) eg:`kbcli addon index update kubeblocks`
4. list: list all addon index eg: `kbcli addon index list`

image

### CODE REVIEW

File: pkg/util/gitutil/git.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
+package gitutil
```
Comment (ldming) [inline pkg/util/gitutil/git.go:1]: why not use `util/git.go`?

File: pkg/cmd/addon/addon.go [hunk 2]
```diff
@@ -100,17 +100,17 @@ func NewAddonCmd(f cmdutil.Factory, streams genericiooptions.IOStreams) *cobra.C
...
 func newListCmd(f cmdutil.Factory, streams genericiooptions.IOStreams) *cobra.Command {
 	o := action.NewListOptions(f, streams, types.AddonGVR())
 	cmd := &cobra.Command{
-		Use:               "list",
-		Short:             "List addons.",
-		Aliases:           []string{"ls"},
-		ValidArgsFunction: util.ResourceNameCompletionFunc(f, o.GVR),
+		Use:     "list",
+		Short:   "List addons.",
+		Aliases: []string{"ls"},
```
Comment (ldming) [inline pkg/cmd/addon/addon.go:113]: Do not modify this command.

File: pkg/cmd/addon/index.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
+package addon
```
Comment (ldming) [inline pkg/cmd/addon/index.go:1]: add license header.

File: pkg/cmd/addon/index.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	"path"
+	"regexp"
+
+	"github.com/apecloud/kbcli/pkg/printer"
+	"github.com/apecloud/kbcli/pkg/types"
+	"github.com/apecloud/kbcli/pkg/util"
+	"github.com/apecloud/kbcli/pkg/util/gitutil"
+	"github.com/spf13/cobra"
+	"k8s.io/cli-runtime/pkg/genericiooptions"
+	"k8s.io/klog/v2"
```
Comment (ldming) [inline pkg/cmd/addon/index.go:17]: adjust import order.

File: pkg/cmd/addon/index.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+func init() {
+	addonDir, err := util.GetCliAddonDir()
+	if err != nil {
+		klog.V(1).ErrorS(err, "can't get the addon ")
+	}
+
+	defaultIndexDir := path.Join(addonDir, types.KubeBlocksReleaseName)
+	if _, err := os.Stat(defaultIndexDir); os.IsNotExist(err) {
+		err = gitutil.EnsureCloned(types.DefaultAddonIndexURL, defaultIndexDir)
```
Comment (ldming) [inline pkg/cmd/addon/index.go:33]: Clone github repo in `init` function is not a good choice, better to clone when run the corresponding command.

File: pkg/cmd/addon/index.go [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+func newIndexAddCmd() *cobra.Command {
+	indexAddCmd := &cobra.Command{
+		Use:     "add",
+		Short:   "Add a new index",
+		Long:    "Configure a new index to install KubeBlocks addon from.",
+		Example: "kbcli index add KubeBlocks " + types.DefaultAddonIndexURL,
+		Args:    cobra.ExactArgs(2),
+		Run: func(_ *cobra.Command, args []string) {
+			util.CheckErr(addAddonIndex(args))
```
Comment (ldming) [inline pkg/cmd/addon/index.go:65]: named `addIndex(args)`

File: pkg/cmd/addon/index.go [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+			fmt.Printf("You have added a new index from %q\nThe addons in this index are not audited by ApeCloud.", args[1])
```
Comment (ldming) [inline pkg/cmd/addon/index.go:66]: Is "The addons in this index are not audited by ApeCloud." needed?

File: pkg/cmd/addon/index.go [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+		},
+	}
+	return indexAddCmd
+}
+
+func newIndexDeleteCmd() *cobra.Command {
+	indexDeleteCmd := &cobra.Command{
+		Use:   "remove",
```
Comment (ldming) [inline pkg/cmd/addon/index.go:74]: Why `remove`, not `delete`?

File: pkg/cmd/addon/index.go [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+		Short: "Remove an addon index",
+		Long:  `Remove a configured addon index.`,
+
+		Args: cobra.ExactArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
+			util.CheckErr(indexDelete(args[0]))
```
Comment (ldming) [inline pkg/cmd/addon/index.go:80]: `deleteIndex(..`

File: pkg/cmd/addon/index.go [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+		},
+	}
+
+	return indexDeleteCmd
+}
+
+func newIndexCmd(streams genericiooptions.IOStreams) *cobra.Command {
+	indexCmd := &cobra.Command{
+		Use:   "index",
+		Short: "Manage custom plugin indexes",
```
Comment (ldming) [inline pkg/cmd/addon/index.go:90]: modify the Short and Long description, NOT plugin.

File: pkg/cmd/addon/index.go [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+	return indexCmd
+}
+
+// IsValidIndexName validates if an index name contains invalid characters
+func IsValidIndexName(name string) bool {
+	var validNamePattern = regexp.MustCompile(`^[A-Za-z0-9_-]+$`)
+	return validNamePattern.MatchString(name)
+}
+
+func addAddonIndex(args []string) error {
```
Comment (ldming) [inline pkg/cmd/addon/index.go:109]: addIndex

File: pkg/types/types.go [hunk 12]
```diff
@@ -48,6 +48,11 @@ const (
 	// DefaultLogFilePrefix is the default log file prefix
 	DefaultLogFilePrefix = "kbcli"
 
+	// AddonIndexEnv defines kbcli addon index dir
+	AddonIndexEnv = "KBCLI_ADDON_INDEX_DIR"
+	// AddonIndexDir is the default addon index dir
+	AddonIndexDir = "addon_index"
```
Comment (ldming) [inline pkg/types/types.go:54]: Suggest the addon index path is `$CLIHOME/addons/index`

File: pkg/cmd/addon/index.go [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+func newIndexAddCmd() *cobra.Command {
+	indexAddCmd := &cobra.Command{
+		Use:     "add",
+		Short:   "Add a new addon index",
+		Long:    "Configure a new index to install KubeBlocks addon from.",
+		Example: "kbcli index add kubeblocks " + types.DefaultAddonIndexURL,
+		Args:    cobra.ExactArgs(2),
+		Run: func(_ *cobra.Command, args []string) {
+			util.CheckErr(addIndex(args))
+			fmt.Printf("You have added a new index from %q\n", args[1])
```
Comment (ldming) [inline pkg/cmd/addon/index.go:68]: Print info in the addIndex function, check other similar.

File: pkg/cmd/addon/index.go [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+		},
+	}
+
+	defaultIndexCmd := &cobra.Command{
```
Comment (ldming) [inline pkg/cmd/addon/index.go:72]: Remove this command, and like https://github.com/apecloud/kbcli/blob/29ffeb26dd7cc2707c1c5024464c054000fcb4f5/pkg/cmd/plugin/plugin.go#L61-L61, when run corresponding subcommand, check index and init the default index if not set.

File: pkg/cmd/addon/index.go [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+	}
+	return nil
+}
+
+func (o *updateOption) run() error {
+	addonDir, err := util.GetCliAddonDir()
+	if err != nil {
+		return err
+	}
+	// update := []string{}
```
Comment (ldming) [inline pkg/cmd/addon/index.go:160]: remove it

File: pkg/types/types.go [hunk 16]
```diff
@@ -51,6 +51,11 @@ const (
 	// DefaultLogFilePrefix is the default log file prefix
 	DefaultLogFilePrefix = "kbcli"
 
+	// AddonIndexEnv defines kbcli addon index dir
+	AddonIndexEnv = "KBCLI_ADDON_INDEX_DIR"
+	// AddonIndexDir is the default addon index dir
+	AddonIndexDir = "index"
```
Comment (ldming) [inline pkg/types/types.go:57]: I prefer to use `addons/index` directory.

File: pkg/cmd/addon/index.go [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+	}
+	return indexListCmd
+}
+
+func newIndexAddCmd() *cobra.Command {
+	indexAddCmd := &cobra.Command{
+		Use:     "add",
+		Short:   "Add a new addon index",
+		Long:    "Configure a new index to install KubeBlocks addon from.",
+		Example: "kbcli index add kubeblocks " + types.DefaultAddonIndexURL,
```
Comment (ldming) [inline pkg/cmd/addon/index.go:64]: This example should be `kbcli addon index add kubeblocks ...`

File: pkg/cmd/addon/index.go [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
+	o := &updateOption{
+		names:     make([]string, 0),
+		all:       false,
+		IOStreams: streams,
+	}
+	indexUpdateCmd := &cobra.Command{
+		Use:               "update",
+		Short:             "update the specified index(es)",
+		Long:              "Update existed index repository from index origin URL",
+		Example:           "kbcli index update KubeBlocks",
```
Comment (ldming) [inline pkg/cmd/addon/index.go:177]: `kbcli addon index update KubeBlocks`

File: docs/user_docs/cli/kbcli_addon.md [hunk 19]
```diff
@@ -40,6 +40,7 @@ Addon command.
 * [kbcli addon describe](kbcli_addon_describe.md)	 - Describe an addon specification.
 * [kbcli addon disable](kbcli_addon_disable.md)	 - Disable an addon.
 * [kbcli addon enable](kbcli_addon_enable.md)	 - Enable an addon.
+* [kbcli addon index](kbcli_addon_index.md)	 - Manage custom addon indexes
```
Comment (ldming) [inline docs/user_docs/cli/kbcli_addon.md:43]: Where is the `kbcli_addon_index.md`

Comment (1aal) [reply docs/user_docs/cli/kbcli_addon.md:43]: > Where is the `kbcli_addon_index.md`

Due to the wrong usage, the user documentation was not generated automatically. I have made the adjustments.

**Review by ldming — APPROVED** [review]

**Pull request merged**